### PR TITLE
Implement/change/expose to Lua constructions findAtTile & insert

### LINKED ISSUE #2323: Convert Constructions::findAtTile(df::coord pos) to binary search and expose to Lua

### PR DESCRIPTION
**Pull request by wolfboyft**
Fixes #2323

### CODE REVIEW

File: docs/Lua API.rst [hunk 1]
```diff
@@ -1906,6 +1906,13 @@ Constructions module
   coordinates, designates it for removal, or instantly cancels the planned one.
   Returns *true, was_only_planned* if removed; or *false* if none found.
 
+* ``dfhack.constructions.findAtTile(pos)``, or ``findAtTile(x,y,z)``
+
+  Returns the construction at the given position, or ``nil`` if there isn't one.
+
+* ``dfhack.constructions.insert(construction)``
+
+  Properly inserts the given construction into the game.
```
Comment (myk002) [inline docs/Lua API.rst:1915]: be sure to note that it returns a boolean that indicates successful insertion.

File: library/LuaApi.cpp [hunk 2]
```diff
@@ -2276,8 +2276,17 @@ static int constructions_designateRemove(lua_State *L)
...
+{
+    auto pos = CheckCoordXYZ(L, 1, true);
+    Lua::PushDFObject(L, Constructions::findAtTile(pos));
+    return 1;
+}
+
 static const luaL_Reg dfhack_constructions_funcs[] = {
     { "designateRemove", constructions_designateRemove },
+    { "findAtTile", constructions_findAtTile },
+    WRAPM(Constructions, insert),
```
Comment (myk002) [inline library/LuaApi.cpp:2289]: the `WRAPM` line should go in the `dfhack_constructions_module` array on line 2265

File: library/modules/Constructions.cpp [hunk 3]
```diff
@@ -54,11 +54,18 @@ using df::global::world;
...
-            return *it;
+    int index = binsearch_index(world->constructions, pos);
+    if (index == -1) {
+        return NULL;
     }
-    return NULL;
+    return world->constructions[index];
+}
+
+bool Constructions::insert(df::construction * constr)
```
Comment (myk002) [inline library/modules/Constructions.cpp:64]: this needs a prototype in the `library/include/modules/Constructions.h` file

**Review by myk002 — CHANGES_REQUESTED** [review]

**Comment by wolfboyft** [comment]
What? Errors? I just built it myself!

**Comment by myk002** [comment]
For the final touch, could you change the PR description to say "Fixes #2323" so the related issue gets closed when we merge this PR? Example: https://github.com/DFHack/dfhack/pull/2318

**Comment by myk002** [comment]
> Wait, don't we also need a changelog entry?

Ha, yes, yes we do.

**Comment by wolfboyft** [comment]
Misc improvements for the binsearch, API for the new function, and Lua for the two newly exposed functions?

**Comment by myk002** [comment]
Sgtm

**Comment by wolfboyft** [comment]
Let me know if there're any gripes with the changelog entries.

**Comment by wolfboyft** [comment]
It's strange to get new compile errors when all I changed was the changelog...

**Review by myk002 — APPROVED** [review]

**Pull request merged**